Log only a summary line of filesystem detail for all path.data on node startup

mikemccand opened the pull request:
This is a follow-on from #10502: I fixed the logging in NodeEnvironment so that if DEBUG is enabled you get (verbose) one line per path.data, e.g.:

```
  [2015-04-10 07:15:16,342][DEBUG][env                      ] [node_t2] node data locations details:
   -> /l/es.fsstats/target/J0/./tests-20150410181516-239/0000 has-space/multi-path1/TEST-haswell-CHILD_VM=[0]-CLUSTER_SEED=[-5088446507314103482]-HASH=[B3A9F4276F5BF]/nodes/0, free_space [263.9gb], usable_space [259.4gb], total_space [465gb], spins? [no], mount [/ (/dev/mapper/haswell--vg-root)], type [btrfs]
   -> /l/es.fsstats/target/J0/./tests-20150410181516-239/0000 has-space/multi-path2/TEST-haswell-CHILD_VM=[0]-CLUSTER_SEED=[-5088446507314103482]-HASH=[B3A9F4276F5BF]/nodes/0, free_space [263.9gb], usable_space [259.4gb], total_space [465gb], spins? [no], mount [/ (/dev/mapper/haswell--vg-root)], type [btrfs]
```

(I also removed the extra newline)

else if INFO is enabled (ES's default) you get a single summary line:

```
  1> [2015-04-10 07:08:48,560][INFO ][env                      ] [node_t2] using [2] data paths, net usable_space [259.4gb], net total_space [465gb], types [btrfs], spins? [no]
```

I also added filesystem type and spins to FsStats.Info, and fixed JmxFsProbe to set them and to also set mount (I use FileStore.toString() for this). 

For SigarFsType, I used its FileSystem.getSysTypeName to get the filesystem type (not certain this is correct!), and also call Lucene's IOUtils.spins.

On backport to 1.x I'll fix the readFrom/writeTo to be back-compat if version < 1.6.

rmuir commented:
> I also added filesystem type and spins to FsStats.Info, and fixed JmxFsProbe to set them and to also set mount (I use FileStore.toString() for this).

when does this get called? These operations should not be assumed to be fast.

mikemccand commented:
> when does this get called? These operations should not be assumed to be fast.

Hmm good point: they are called once per second (by default).

Maybe ... they should not be in FsProbe?  Maybe it should be in FsService.stats, so it's only done when stats are requested?  Or maybe we call this once and cache (in NodeEnv)?

rmuir commented:
IMO the code in question (IOUtils.spins) is too scary to call once per second. Today its only called once per writer on the first merge. It should be cached.

mikemccand commented:
I changed NodeEnvironment to cache the FileStore and IOUtils.spins result, in a new static NodePath class.

I left the separate Lock[] as a private parallel array: I didn't want callers to have access to the lock when they get the NodePath[].

Note that this made NodeEnv.nodeDataPaths and .indexPaths methods a bit more costly since they now build a new array on each call, but I reviewed places that call it and I don't think the added cost will matter.  If that's a problem I could pre-compute and cache that too.

Also, I noticed dead code in NodeEnv (findAllShardIds) and removed it.

mikemccand commented:
@rmuir any feedback here?

rmuir commented on src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -50,13 +52,35 @@
+        public final Path indicesPath;
+        /** Cached FileStore from path */
+        public final FileStore fileStore;
+        /** Cached result of Lucene's {@code uIOUtils.spins} on path */
```
can we fix the typo and also state that it is a trilean?

rmuir commented on src/main/java/org/elasticsearch/env/NodeEnvironment.java:
```diff
@@ -143,60 +160,123 @@ public NodeEnvironment(Settings settings, Environment environment) throws IOExce
     }
 
+
+    // TODO: move somewhere more "util"?  But, this is somewhat hacky code ... not great to publicize it any more:
```
can the methods like getFileStore() be package private? I don't see anything outside of this file using them.

rmuir commented:
i added some minor comments. otherwise, looks good here, thanks for caching this.
